base/array.jl

```diff
@@ -237,16 +237,29 @@ function _collect(cont, itr, ::HasEltype, isz::SizeUnknown)
     return a
 end
 
+if isdefined(:Base)
+@eval function _default_eltype(itrt::ANY)
+    $(Expr(:meta, :pure))
+    try
+        rt = return_types(first, (itrt,))
```

Comment:
Yeah, it's tricky. Part of the difference between this and static_typeof is that this use is expected to be best-effort: we don't need *the* inferred type, just whatever type we can get, and if it's not a leaf type we ignore it anyway. Or we can just throw an error in the empty case.